### ISSUE TITLE
5MB/minute memory leak with no extensions

### ISSUE THREAD

**Issue by officiallymarky** [microsoft/vscode#156777]
Issue Type: <b>Bug</b>

I have 45 extensions installed and applications start to crash after leaving VS Code running for 2-3 days, if I have more than one instance it is sooner.

I assumed it was an extension as I have 45 installed.  My initial reaction was to disable GitHub CoPilot thinking this might be the problem, and no change in memory loss.  I am using VS Code Process Explorer to monitor memory usage.

So my next step was to disable all extensions and enable them one by one to see if I can catch the one giving me a problem.  While waiting a few seconds for a baseline with no extensions, I noticed I was still losing about 5MB/minute or just over 300Mb/hour.

I am using latest version on Windows 10 Pro x64 with latest system updates.  I have tried a fresh reboot and running without extensions on a fresh restart and I am continously losing ram at the same rate.  With all extensions enabled, I am losing roughly the same amount of ram per minute.

I've been having this issue for a few months and doubled my system ram to 64G to give me more buffer until things start crashing as at the time I haven't narrowed down the problem.

UPDATE:
I installed the insider build and seeing similar results, in fact memory loss is closer to 5.9MB/minute.  This is without extensions and with default settings file.  Just sitting idle without a project loaded.

VS Code version: Code 1.69.2 (3b889b090b5ad5793f524b5d1d39fda662b96a2a, 2022-07-18T16:12:52.460Z)
OS version: Windows_NT x64 10.0.19044
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 9 5950X 16-Core Processor             (32 x 3400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|63.92GB (43.29GB free)|
|Process Argv|--crash-reporter-id 5e70ba6f-248a-4f0a-8205-dd6e93c2c8dc|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (43)</summary>

Extension|Author (truncated)|Version
---|---|---
markdown-mermaid|bie|1.14.2
vscode-tailwindcss|bra|0.8.6
npm-intellisense|chr|1.4.2
postcss|css|1.0.9
transformer|dak|1.12.1
vscode-eslint|dba|2.2.6
githistory|don|0.6.19
vscode-babel-coloring|dza|0.0.4
gitlens|eam|12.1.2
vscode-html-css|ecm|1.13.1
EditorConfig|Edi|0.16.4
vscode-npm-script|eg2|0.3.28
prettier-vscode|esb|9.5.0
text-transform|flo|0.1.0
auto-rename-tag|for|0.1.10
copilot|Git|1.36.6340
todo-tree|Gru|0.0.215
dotenv|mik|1.0.1
inline-fold|moa|0.1.6
python|ms-|2022.10.1
vscode-pylance|ms-|2022.7.40
jupyter|ms-|2022.6.1201981810
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.9
remote-ssh|ms-|0.84.0
remote-ssh-edit|ms-|0.80.0
remote-wsl|ms-|0.66.3
color-highlight|nau|2.5.0
indent-rainbow|ode|8.3.1
hivecoin-statusbar|Off|0.0.5
quicktype|qui|12.0.46
vscode-thunder-client|ran|1.16.7
ansible|red|0.11.10
vscode-yaml|red|1.9.1
wordcount|rid|0.1.33
vue-vscode-snippets|sdr|3.1.1
autoimport|ste|1.5.4
es6-string-html|Tob|2.12.0
highlight-matching-tag|vin|0.10.1
vscode-icons|vsc|11.13.0
volar|Vue|0.39.3
vue|Wsc|1.0.26
markdown-all-in-one|yzh|3.4.3

(2 theme extensions excluded)

</details>

**Comment by rebornix** [comment]
@officiallymarky thanks for the report. From the process explorer, do you see which process is consuming more and more memory? A screenshot would probably help and we can check if it's a UI process or an extension process.

**Comment by officiallymarky** [comment]
> @officiallymarky thanks for the report. From the process explorer, do you see which process is consuming more and more memory? A screenshot would probably help and we can check if it's a UI process or an extension process.

I disabled all extensions, and even ran insider with default config and no extensions.  

It shows all the memory on the code parent process, but doesn't seem to show which one.  

Here is a blank project in Code Insiders with default settings and no extensions before and after a couple of hours (~3 or so)

![explorer_EPJtOl7aHW](https://user-images.githubusercontent.com/33408375/182256030-40373332-470d-42d6-8675-3c27b8b25d1d.png)

![explorer_LB7AqTAAJm](https://user-images.githubusercontent.com/33408375/182256037-63d018a6-74c8-4a5f-9078-fd55a8ed914e.png)

**Comment by rebornix** [comment]
@officiallymarky the screenshots are very helpful, thanks!

**Comment by officiallymarky** [comment]
I still see it with latest update, also I see it in the insiders build.  Has something been identified that is causing this?

**Comment by rzhao271** [comment]
I found a probable memory leak in some code for the process explorer itself. I hope to submit a PR for that later this week. Meanwhile some questions:
1. Does the issue occur for you when you close the process explorer and reopen it after a long while?
2. Does the Windows task manager process explorer report ever increasing numbers for VS Code's memory consumption?

**Comment by officiallymarky** [comment]
> I found a probable memory leak in some code for the process explorer itself. I hope to submit a PR for that later this week. Meanwhile some questions:
> 
> 1. Does the issue occur for you when you close the process explorer and reopen it after a long while?
> 2. Does the Windows task manager process explorer report ever increasing numbers for VS Code's memory consumption?

1. Quick check (5-10 mins) it seems the memory leak slows down/pauses when process explorer isn't running.
2. Yes, but slows down with process explorer is closed

I am really confused, because I never used process explorer until I had this problem.  In fact, I had to buy another 32G of ram (64G total just to run VS Code) as my computer would keep crashing every two days due to this issue.  I got about 2 days before my 32G would wipe out and apps would start crashing.  Took me a while to narrow it down to VS Code.  I keep my computer running and I usually keep one or two VS Code sessions open.  I was seeing same behavior in both VS Code and VS Code Insiders.

I did some testing, and here is what I see.

If I start VS Code, and start Process Explorer, the ram for the main process goes up about 5MB/minute.  I see a similar gain in Task Manager (Windows 10).

If I stop Process Explorer while taking note of the memory in both Process Explorer and Task Manager I notice that Process Manager's memory no longer increments, the next time I open it it is where it was when I closed it, but nothing is freed.  Task Manager though shows some memory releasing, slowly, but not a lot.  For example after a few minutes I am at 340MB in Task Manager and 190 MB as the top process in Process Explorer.  After 10 minutes I may drop 10-20MB in Task Manager but will be at the exact same ram in Process Manager (at least for the first process which is what I am monitoring and what is growing 5MB/minute).  Process Manager will immediately start gaining memory the moment I open it.

This feels like different behavior than before, because in the past my committed memory would constantly grow no matter how much ram I had.  I went from 32G ram to 64G with 35G swap file and I would still exhaust it.  I have since started closing VS Code as soon as I am done with anything as I can't afford to have applications crashing and losing data.  So I don't really know the current state of things outside of this test.  I was losing important information and processes as they would just randomly lock up once I filled committed memory despite having way more than ram that I can possibly need.
